Add conversion-verifier to CI

Populating this issue from https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/3085 to track conversion-verifier addition separately.

```
Integrate the conversion-verifier in CI.

The conversion-verifier can be found here:

Code: https://github.com/kubernetes-sigs/cluster-api/tree/main/hack/tools/conversion-verifier
Usage: https://github.com/kubernetes-sigs/cluster-api/blob/main/Makefile#L339-L341 (core CAPI runs it as part of the verify job)

```

/good-first-issue
/priority backlog

Hey @sedefsavas, I've made some changes locally but I would like to know the appropriate way in which I could test them. Can you please tell me how I can verify if the changes would work properly once implemented in the repo?

I tried to run `make verify` on my local system, but it seems to throw the following error:
```logtalk
❯ make verify
rm ./hack/tools/bin/*.sh
rm: ./hack/tools/bin/*.sh: No such file or directory
make: [verify-boilerplate] Error 1 (ignored)
./hack/verify-boilerplate.sh

KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
dirname "${BASH_SOURCE[0]}"

boilerDir="${KUBE_ROOT}/hack/boilerplate"
boiler="${boilerDir}/boilerplate.py"

files_need_boilerplate=()
while IFS=$'\n' read -r line; do
  files_need_boilerplate+=( "$line" )
done < <("${boiler}" "$@")
"${boiler}" "$@"

# Run boilerplate check
if [[ ${#files_need_boilerplate[@]} -gt 0 ]]; then
  for file in "${files_need_boilerplate[@]}"; do
    echo "Boilerplate header is wrong for: ${file}" >&2
  done

  exit 1
fi
Boilerplate header is wrong for: /Users/abhinav/git-forks/capa/cmd/clusterawsadm/ami/helper_test.go
make: *** [verify-boilerplate] Error 1
```
which makes me think that either there must be some steps prior to executing this command that I'm missing out, or that it's only supposed to be executed in the pipeline.

Any help would be really appreciated, thanks!

--- a/Makefile
+++ b/Makefile
@@ -47,6 +47,7 @@ GO_APIDIFF := $(TOOLS_BIN_DIR)/go-apidiff
 CLUSTERCTL := $(BIN_DIR)/clusterctl
 CONTROLLER_GEN := $(TOOLS_BIN_DIR)/controller-gen
 CONVERSION_GEN := $(TOOLS_BIN_DIR)/conversion-gen
+CONVERSION_VERIFIER := $(TOOLS_BIN_DIR)/conversion-verifier
 DEFAULTER_GEN := $(TOOLS_BIN_DIR)/defaulter-gen
 ENVSUBST := $(TOOLS_BIN_DIR)/envsubst
 GH := $(TOOLS_BIN_DIR)/gh
@@ -590,7 +591,7 @@ clean-release: ## Remove the release folder
 	rm -rf $(RELEASE_DIR)
 
 .PHONY: verify
-verify: verify-boilerplate verify-modules verify-gen release-manifests
+verify: verify-boilerplate verify-modules verify-gen verify-conversions release-manifests
 
 .PHONY: verify-boilerplate
 verify-boilerplate:
@@ -603,6 +604,11 @@ verify-modules: modules
 		git diff; \
 		echo "go module files are out of date"; exit 1; \
 	fi
+
+.PHONY: verify-conversions
+verify-conversions: $(CONVERSION_VERIFIER)  ## Verifies expected API conversion are in place
+	echo verification of api conversions initiated
+	$(CONVERSION_VERIFIER)
 
 verify-gen: generate
 	@if !(git diff --quiet HEAD); then \

--- a/api/v1alpha4/awscluster_conversion.go
+++ b/api/v1alpha4/awscluster_conversion.go
@@ -70,45 +70,6 @@ func (r *AWSCluster) ConvertFrom(srcRaw conversion.Hub) error {
 	return nil
 }
 
-// ConvertTo converts the v1alpha3 AWSCluster receiver to a v1beta1 AWSCluster.
-func (r *AWSClusterTemplate) ConvertTo(dstRaw conversion.Hub) error {
-	dst := dstRaw.(*infrav1.AWSClusterTemplate)
-
-	if err := Convert_v1alpha4_AWSClusterTemplate_To_v1beta1_AWSClusterTemplate(r, dst, nil); err != nil {
-		return err
-	}
-
-	// Manually restore data.
-	restored := &infrav1.AWSClusterTemplate{}
-	if ok, err := utilconversion.UnmarshalData(r, restored); err != nil || !ok {
-		return err
-	}
-
-	dst.Spec.Template.ObjectMeta = restored.Spec.Template.ObjectMeta
-
-	return nil
-}
-
-// ConvertFrom converts the v1beta1 AWSCluster receiver to a v1alpha3 AWSCluster.
-func (r *AWSClusterTemplate) ConvertFrom(srcRaw conversion.Hub) error {
-	src := srcRaw.(*infrav1.AWSClusterTemplate)
-
-	if err := Convert_v1beta1_AWSClusterTemplate_To_v1alpha4_AWSClusterTemplate(src, r, nil); err != nil {
-		return err
-	}
-
-	// Preserve Hub data on down-conversion.
-	if err := utilconversion.MarshalData(src, r); err != nil {
-		return err
-	}
-
-	return nil
-}
-
-func Convert_v1beta1_AWSClusterTemplateResource_To_v1alpha4_AWSClusterTemplateResource(in *infrav1.AWSClusterTemplateResource, out *AWSClusterTemplateResource, s apiconversion.Scope) error {
-	return autoConvert_v1beta1_AWSClusterTemplateResource_To_v1alpha4_AWSClusterTemplateResource(in, out, s)
-}
-
 // Convert_v1alpha4_APIEndpoint_To_v1beta1_APIEndpoint .
 func Convert_v1alpha4_APIEndpoint_To_v1beta1_APIEndpoint(in *clusterv1alpha4.APIEndpoint, out *clusterv1.APIEndpoint, s apiconversion.Scope) error {
 	return clusterv1alpha4.Convert_v1alpha4_APIEndpoint_To_v1beta1_APIEndpoint(in, out, s)

--- a/api/v1alpha4/awsclustertemplate_conversion.go
+++ b/api/v1alpha4/awsclustertemplate_conversion.go
@@ -1,0 +1,85 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha4
+
+import (
+	apiconversion "k8s.io/apimachinery/pkg/conversion"
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1beta1"
+	utilconversion "sigs.k8s.io/cluster-api/util/conversion"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+)
+
+// ConvertTo converts the v1alpha4 AWSClusterTemplate receiver to a v1beta1 AWSClusterTemplate.
+func (r *AWSClusterTemplate) ConvertTo(dstRaw conversion.Hub) error {
+	dst := dstRaw.(*infrav1.AWSClusterTemplate)
+
+	if err := Convert_v1alpha4_AWSClusterTemplate_To_v1beta1_AWSClusterTemplate(r, dst, nil); err != nil {
+		return err
+	}
+
+	// Manually restore data.
+	restored := &infrav1.AWSClusterTemplate{}
+	if ok, err := utilconversion.UnmarshalData(r, restored); err != nil || !ok {
+		return err
+	}
+
+	dst.Spec.Template.ObjectMeta = restored.Spec.Template.ObjectMeta
+
+	return nil
+}
+
+// ConvertFrom converts the v1beta1 AWSClusterTemplate receiver to a v1alpha4 AWSClusterTemplate.
+func (r *AWSClusterTemplate) ConvertFrom(srcRaw conversion.Hub) error {
+	src := srcRaw.(*infrav1.AWSClusterTemplate)
+
+	if err := Convert_v1beta1_AWSClusterTemplate_To_v1alpha4_AWSClusterTemplate(src, r, nil); err != nil {
+		return err
+	}
+
+	// Preserve Hub data on down-conversion.
+	if err := utilconversion.MarshalData(src, r); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ConvertTo converts the v1alpha4 AWSClusterTemplateList receiver to a v1beta1 AWSClusterTemplateList.
+func (r *AWSClusterTemplateList) ConvertTo(dstRaw conversion.Hub) error {
+	dst := dstRaw.(*infrav1.AWSClusterTemplateList)
+
+	if err := Convert_v1alpha4_AWSClusterTemplateList_To_v1beta1_AWSClusterTemplateList(r, dst, nil); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ConvertFrom converts the v1beta1 AWSClusterTemplateList receiver to a v1alpha4 AWSClusterTemplateList.
+func (r *AWSClusterTemplateList) ConvertFrom(srcRaw conversion.Hub) error {
+	src := srcRaw.(*infrav1.AWSClusterTemplateList)
+
+	if err := Convert_v1beta1_AWSClusterTemplateList_To_v1alpha4_AWSClusterTemplateList(src, r, nil); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func Convert_v1beta1_AWSClusterTemplateResource_To_v1alpha4_AWSClusterTemplateResource(in *infrav1.AWSClusterTemplateResource, out *AWSClusterTemplateResource, s apiconversion.Scope) error {
+	return autoConvert_v1beta1_AWSClusterTemplateResource_To_v1alpha4_AWSClusterTemplateResource(in, out, s)
+}

--- a/api/v1beta1/conversion.go
+++ b/api/v1beta1/conversion.go
@@ -54,3 +54,6 @@ func (*AWSClusterControllerIdentityList) Hub() {}
 
 // Hub marks AWSClusterTemplate as a conversion hub.
 func (*AWSClusterTemplate) Hub() {}
+
+// Hub marks AWSClusterTemplateList as a conversion hub.
+func (*AWSClusterTemplateList) Hub() {}

--- a/cmd/clusterawsadm/ami/helper_test.go
+++ b/cmd/clusterawsadm/ami/helper_test.go
@@ -1,3 +1,19 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
 package ami
 
 import (

--- a/hack/tools/Makefile
+++ b/hack/tools/Makefile
@@ -76,6 +76,10 @@ CONVERSION_GEN := $(BIN_DIR)/conversion-gen
 $(CONVERSION_GEN): $(BIN_DIR) go.mod go.sum $(call rwildcard,.,third_party/conversion-gen/*.*)
 	go build -tags=tools -o $@ ./third_party/conversion-gen
 
+CONVERSION_VERIFIER := $(BIN_DIR)/conversion-verifier
+$(CONVERSION_VERIFIER): $(BIN_DIR) go.mod go.sum
+	go build -tags=tools -o $@ sigs.k8s.io/cluster-api/hack/tools/conversion-verifier
+
 DEFAULTER_GEN := $(BIN_DIR)/defaulter-gen
 $(DEFAULTER_GEN): $(BIN_DIR) go.mod go.sum
 	go build -tags=tools -o $@ k8s.io/code-generator/cmd/defaulter-gen

--- a/hack/tools/tools.go
+++ b/hack/tools/tools.go
@@ -1,3 +1,4 @@
+//go:build tools
 // +build tools
 
 /*
@@ -21,6 +22,7 @@ package tools
 
 import (
 	_ "embed"
+
 	_ "github.com/a8m/envsubst"
 	_ "github.com/ahmetb/gen-crd-api-reference-docs"
 	_ "github.com/golang/mock/mockgen"
@@ -31,6 +33,7 @@ import (
 	_ "k8s.io/code-generator"
 	_ "k8s.io/code-generator/cmd/conversion-gen"
 	_ "k8s.io/release/cmd/release-notes"
+	_ "sigs.k8s.io/cluster-api/hack/tools/conversion-verifier"
 	_ "sigs.k8s.io/cluster-api/hack/tools/mdbook/embed"
 	_ "sigs.k8s.io/controller-tools/cmd/controller-gen"
 	_ "sigs.k8s.io/kind"

